design/authorization-handler.md

```diff
@@ -0,0 +1,590 @@
+ 
+So many considerations, benefits and implementation details of both proposals can be considered true here, as well as some of the ideas proposed in the [Dynamic Scopes vs RAR discussion](https://github.com/keycloak/keycloak/discussions/8488).
+
+## Authorization Handler VS Client scope
+
+In this design, we are introducing a new entity called Authorization Handler, which is supposed to replace existing Client Scope. There are two possible views how these changes can be seen:
+
+* Introducing new entity "Authorization Handler", which is replacing "Client scope"
+* Extending existing "Client Scope" entity with more capabilities described in this design (dynamic scopes and RAR requests in addition to just “static scope” requests supported by client scopes in current Keycloak)
+
+The overall output should be the same in both cases. From the implementation and backwards compatibility perspective, it might be probably easier to re-use the existing Client Scope entity under the covers, but we are not 100% sure (this will require us to take a look at the current implementation to decide about the best path).
+
+Regarding migration and backwards compatibility, we will need to make sure that existing Client Scopes would be migrated to the Authorization Handler. So users migrating from Keycloak 15.0.2, who have set up some Client Scope will still have exactly the same functionality in new Keycloak versions as before.
+
+In the next text, it can be assumed that Authorization Handler and Client Scope are the same entity for all purposes. So by Client Scope is meant the "Dynamic" client scope, which will be able to handle dynamic scopes and RAR requests. Not just "Static client scope" as available in Keycloak in version 15.0.2.
+
+**Note**: We need community feedback on what the best path would be, and if we decide to go with the authorization handler, it would also be great to get a better name for it.
+
+## The Authorization Handler
+
+A new type of entity will be created in Keycloak that will allow users to configure the behaviour of Static, Dynamic Scopes and RAR in a composable way.
+
+The AuthorizationHandler will be a composite structure consisting of the following pluggable components:
+
+* The AuthorizationContext initiator
+* A Compliance processor
+* A consent screen enhancer / builder
+* Configured Protocol Mappers.
+
+
+### AuthorizationContext initiator
+
+At the initial step, the authorization data sent by the client can be retrieved from the initial OIDC Authorization-Endpoint request (or Token-Request in the use-cases like CIBA or OAuth2 resource-owner-password-credentials grant). This retrieved data will be referred to as AuthorizationRequest, which means the initial request to Keycloak to start the authentication process.
```

Comment:
@mposolda +1

Reply:
I was also confused and I @mposolda helps to makt it more clear. But isn't it just saying that "Access/Authorization data is represented by a `AuthorizationRequest`"? So that no matter what grant type / request mode you are using it means the access/authorization data sent along with an authorization request?

Reply:
Yes, probably the "initial OIDC Authorization-Endpoint request" can be simplified to "Authorization Request" to make it shorter in the text?

But IMO it will be clear to still include this NOTE and mention the various type of requests we're talking about (EG. BackchannelAuthenticationEndpoint in case of CIBA etc).